stm32/isr: Add option to run flash & uart isr function from ram.

andrewleech opened the pull request:
This allows uart rx to function while flash erase/writes operations
are under way, preventing lost data so long as it fits in the uart
rx buffer.

This is a more general implementation of https://github.com/micropython/micropython/pull/8713 which is intended to fix https://github.com/micropython/micropython/issues/8386

Enable in mpconfigboard.mk:
MICROPY_USE_RAM_ISR_UART_FLASH_FN = 1

It's been enabled on the NUCLEO_WB55 but should equally apply to any board with its default repl on a uart port without flow control (most nucleo boards with onboard stlink).

This PR is currently sitting on top of / includes https://github.com/micropython/micropython/pull/9053 which makes the diff somewhat messy to view on github unfortunately.

It also includes a commit to pre-process linker scripts.
This allows the use of command line -D defines and/or including header files to switch features on/off in the linker scripts. 

It's used in the `ld` files to select between `common_isr.ld` and `common_isr_ram.ld` snippet.

pi-anl committed:
stm32: Add option to put ISR, flash and UART code in RAM.

This allows UART RX to function while flash erase/writes operations are
under way, preventing lost serial data so long as it fits in the UART RX
buffer.

This enables (among other things) mpremote to successfully copy files to
boards that use a UART REPL.

Enable via the following option placed in `mpconfigboard.mk`:

    MICROPY_HW_ENABLE_ISR_UART_FLASH_FUNCS_IN_RAM = 1

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

dlech commented:
Doesn't this require that *all* ISR functions and their depedencies be moved to RAM? Or does this somehow rely on interrupt priorities or something like that to ensure that only certain interrupts are enabled during flashing. For example, what happens if someone implements a GPIO interrupt handler in Python and that is triggered while flashing?

andrewleech commented:
@dlech yes it's a good question that I'm still not 100% certain about. The irq priorities might be enough protection, that's why pend and systick are in the list being higher prio. 

It would be good to test with a high frequency timer interrupt I guess to see if that breaks it. If that does lock up we could move the actual flash operation inside an irq at the appropriate priority, that would then definitely block lower prio irqs.

andrewleech commented:
So the pre-processing of linker scripts comes with an undesirable side-effect - lines like `*lib/mbedtls/*(.text* .rodata*)` throw errors because it include `/*` in the middle of the line. While it can be worked around with `*lib/mbedtls/?*(.text* .rodata*)` that's a bit difficult to communicate / document.

As such I've reworked this to PR to no longer use gcc pre-processing of linker files.

FWIW if anyone's interested in the linker preproc it can still be viewed here: https://github.com/micropython/micropython/commit/ffc88fc487ed1a871406582e35a58b5f743a9032

iabdalkader commented:
Hi, should this be enabled by default for every STM board, including ones that don't even use the internal flash storage for the filesystem ?

andrewleech commented:
The latest commit (0abc1d9) has been tested on the USB/UART interface of NUCLEO_WB55 to copy a ~ 66KB file with mpremote. Previously this would fail every time.

andrewleech commented:
> Hi, should this be enabled by default for every STM board, including ones that don't even use the internal flash storage for the filesystem ?

At the moment I believe the only "significant" value of this change is allowing uart repl to work while flash is busy. 

I did question whether having some ISR's in ram would increase performance for all stm's but have heard that the internal flash cache (ART) probably already provides a similar level of performance gain automatically.

andrewleech commented:
TODO re-test whether flash functions need to be in ram; this costs ram in use that may not actually be needed.

andrewleech commented:
Updated and rebased to current master. 

As per https://github.com/micropython/micropython/commit/091aa93ab03a48b8d6947db490691c8863505414#r143545503 I've cleaned up the mechanism for including the correct linker snippet. Re-tested on WB55 to confirm it still works correctly.

I checked through the boards I'd modified and found quite a few that aren't using the internal filesystem, so they've been reverted.

I then rechecked all boards for any new ones that needed it added, updating them as needed.
This following was used for search for all boards which:
* declare `UART_REPL` in mpconfigboard.h
* and do not declare `MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE (0)`
* and do not alreadyt have `MICROPY_USE_RAM_ISR_UART_FLASH_FN ` in mpconfigboard.mk
```
for h in $(grep -l -R UART_REPL ports/stm32/boards/*); do 
grep -q "MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE.*\(0\)" "$h" || \
grep --files-without-match MICROPY_USE_RAM_ISR_UART_FLASH_FN "${h%.*}.mk";
done
```

As per:
> TODO re-test whether flash functions need to be in ram; this costs ram in use that may not actually be needed.

I tested a build removing all the FLASH related from the ram linker listing, found the problem came straight back - copying file causes lockups. Bringing the FLASH functions back into the ram/isr linker fixes it again.

andrewleech commented:
Currently I've removed the FLASH_ISR MEMORY section in linker scripts as the block is now ALIASED into flash or ram as appropriate.
On many (older) boards though, this memory section needs to be limited to 16K as the filesystem is directly after it (to meet sector size needs). 
So without this 16k limited size, if the FLASH_ISR overflows in future it won't get caught by the linker and instead break the filesystem.

So FLASH_ISR (or similar) needs to be added back to the MEMORY section of linker files, at least for boards with a "middle" filesytem, then the ALIAS point to to that instead of just FLASH.

andrewleech commented:
I've updated this to preserve the size-limited first sector in linker where used to ensure that any future overflow is detected correctly.

All boards were compiled to ensure they all _can_ still be compiled:
```
~/micropython $ for board in $(cd ports/stm32/boards; find * -type d); do [ -e ports/stm32/boards/$board/mpconfigboard.mk ] && make -C ports/stm32 BOARD="$board" -j18 BUILD=../../after/build-$board; done
```
and
```
~/micropython $ for board in $(cd ports/stm32/boards; find * -type d); do [ -e ports/stm32/boards/$board/mpconfigboard.mk ] && make -C ports/stm32 BOARD="$board" MBOOT=1 -j18 BUILD=../../after-mboot/build-$board; done
```

dpgeorge commented on ports/stm32/Makefile:
```diff
@@ -143,7 +143,15 @@ CFLAGS += -DMICROPY_FLOAT_IMPL=MICROPY_FLOAT_IMPL_FLOAT
 
-LDFLAGS += -nostdlib -L $(LD_DIR) $(addprefix -T,$(LD_FILES)) -Wl,-Map=$(@:.elf=.map) -Wl,--cref
+# Configure linker include dir for ram/rom isr support
+ifeq ($(MICROPY_USE_RAM_ISR_UART_FLASH_FN),1)
```
How about calling this option `MICROPY_HW_ENABLE_ISR_UART_FLASH_FUNCS_IN_RAM`?

dpgeorge commented on ports/stm32/boards/STM32F769DISC/f769_qspi.ld:
```diff
@@ -38,8 +39,16 @@ _ram_end = ORIGIN(RAM) + LENGTH(RAM);
+_micropy_hw_internal_flash_storage_start = ORIGIN(FLASH_FS);
+_micropy_hw_internal_flash_storage_end = ORIGIN(FLASH_FS) + LENGTH(FLASH_FS);
+
+
```
please only have a single blank line here

dpgeorge commented on ports/stm32/boards/common_isr_ram/common_isr.ld:
```diff
@@ -0,0 +1,40 @@
+    /* These functions need to run from ram to enable uart
+       reception during flash erase/write operations.
+       Defining them here ensures they're copied from
+       flash (in ram_irq.c) along with the isr_vector above.
```
`ram_irq.c` no longer exists

dpgeorge commented on ports/stm32/boards/common_isr_ram/common_isr.ld:
```diff
@@ -0,0 +1,40 @@
+
+} >RAM AT >FLASH_ISR
+
+_start_init_isr = LOADADDR(.isr_vector); /* Used by the start-up code to initialise data */
```
looks like this variable is unused

andrewleech replied:
Ah you're right, seems I had a few different attempts worth of variables defined in here still sorry!

dpgeorge commented on ports/stm32/boards/common_isr_ram/common_isr.ld:
```diff
@@ -0,0 +1,40 @@
+/* Used by the start-up code to initialise data */
+__isr_vector_flash_addr = LOADADDR(.isr_vector);
+__isr_vector_ram_addr = ADDR(.isr_vector);
+__isr_section_len = SIZEOF(.isr_vector);
```
looks like the latter two variables here are unused

dpgeorge commented on ports/stm32/boards/common_isr_rom/common_isr_extratext.ld:
```diff
@@ -0,0 +1,24 @@
+        out.
+
+        NOTE: If you update the list of files contained in .isr_extratext,
+        then be sure to also update smhal/Makefile where it forcibly
```
please replace "smhal" with "stm32"

dpgeorge commented on ports/stm32/boards/common_isr_rom/common_isr_extratext.ld:
```diff
@@ -0,0 +1,24 @@
+.isr_extratext :
+{
+    . = ALIGN(4);
+    /* This first flash block is 16K and the isr vectors only take up
```
suggest adding a blank line between ALIGN and the comment

dpgeorge commented on ports/stm32/boards/common_isr_rom/common_isr_extratext.ld:
```diff
@@ -0,0 +1,24 @@
+        then be sure to also update smhal/Makefile where it forcibly
+        builds each of these files with -Os
+    */
+
```
suggest removing this blank line

andrewleech replied:
Thanks that does make it clearer.

dpgeorge commented on ports/stm32/boards/STM32F769DISC/f769_qspi.ld:
```diff
@@ -38,8 +39,16 @@ _ram_end = ORIGIN(RAM) + LENGTH(RAM);
+
 ENTRY(Reset_Handler)
 
+REGION_ALIAS("FLASH_ISR", FLASH_APP);
```
The addition of this line belongs in the first commit, not the second.

dpgeorge commented:
I tested this on a NUCLEO_WB55 with UART REPL, and it works really well, mpremote cp to the device works.

dpgeorge commented:
I verified that PYBV10 and PYBD_SF6 firmware do not change with this PR.

pi-anl committed:
stm32/boards: Enable RAM_ISR feature on boards with UART REPL.

Allows mpremote file transfer to work correctly when mpremote is used over
the ST-link USB/UART REPL port.

Fixes issue #8386.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

andrewleech commented:
I've updated as suggested thanks @dpgeorge

dpgeorge commented:
Thanks for updating, now merged!
